tt_fabric/mesh_graph.cpp

```diff
@@ -0,0 +1,302 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "mesh_graph.hpp"
+
+#include <fstream>
+#include <iostream>
+
+#include "yaml-cpp/yaml.h"
+
+namespace tt::tt_fabric {
+
+MeshGraph::MeshGraph(const std::string& mesh_graph_desc_file_path) {
+    this->initialize_from_yaml(mesh_graph_desc_file_path);
+}
+
+void MeshGraph::add_to_connectivity(
+    mesh_id_t src_mesh_id,
+    chip_id_t src_chip_id,
+    chip_id_t dest_mesh_id,
+    chip_id_t dest_chip_id,
+    RoutingDirection port_direction) {
+    TT_ASSERT(
+        src_mesh_id < this->intra_mesh_connectivity_.size(),
+        "MeshGraph: Invalid src_mesh_id: {} or unsized intramesh map",
+        src_mesh_id);
+    TT_ASSERT(
+        dest_mesh_id < this->intra_mesh_connectivity_.size(),
+        "MeshGraph: Invalid dest_mesh_id: {} or unsized intramesh map",
+        dest_mesh_id);
+    TT_ASSERT(
+        src_chip_id < this->intra_mesh_connectivity_[src_mesh_id].size(),
+        "MeshGraph: Invalid src_chip_id: {} or unsized intramesh map",
+        src_chip_id);
+    TT_ASSERT(
+        dest_chip_id < this->intra_mesh_connectivity_[dest_mesh_id].size(),
+        "MeshGraph: Invalid dest_chip_id: {} or unsized intramesh map",
+        dest_chip_id);
+
+    TT_ASSERT(
+        src_mesh_id < this->inter_mesh_connectivity_.size(),
+        "MeshGraph: Invalid src_mesh_id: {} or unsized intermesh map",
+        src_mesh_id);
+    TT_ASSERT(
+        dest_mesh_id < this->inter_mesh_connectivity_.size(),
+        "MeshGraph: Invalid dest_mesh_id: {} or unsized intermesh map",
+        dest_mesh_id);
+    TT_ASSERT(
+        src_chip_id < this->inter_mesh_connectivity_[src_mesh_id].size(),
+        "MeshGraph: Invalid src_chip_id: {} or unsized intermesh map",
+        src_chip_id);
+    TT_ASSERT(
+        dest_chip_id < this->inter_mesh_connectivity_[dest_mesh_id].size(),
+        "MeshGraph: Invalid dest_chip_id: {} or unsized intermesh map",
+        dest_chip_id);
+
+    if (src_mesh_id != dest_mesh_id) {
+        // Intermesh Connection
+        auto& edge = this->inter_mesh_connectivity_[src_mesh_id][src_chip_id];
+        if (edge.find(dest_mesh_id) == edge.end()) {
```

Comment:
Here and below: could simplify into with `edge.insert(...)` + handle a case that `dest_mesh_id` already existed there, so you would just manipulate the returned iterator:

```
auto [it, is_inserted] = edge.insert(...);
if (!is_inserted) {
  it->connected_chip_ds.push_back(...);
}
```